README
NAME
    Role::Inspector - introspection for roles

SYNOPSIS
       use strict;
       use warnings;
       use feature qw(say);
   
       {
          package Local::Role;
          use Role::Tiny;   # or Moose::Role, Mouse::Role, etc...
      
          requires qw( foo );
      
          sub bar { ... }
       }
   
       use Role::Inspector qw( get_role_info );
   
       my $info = get_role_info('Local::Role');
   
       say $info->{name};          # Local::Role
       say $info->{type};          # Role::Tiny
       say for @{$info->{api}};    # bar
                                   # foo

DESCRIPTION
    This module allows you to retrieve a hashref of information about a given
    role. The following role implementations are supported:

    *   Moose::Role

    *   Mouse::Role

    *   Moo::Role

    *   Role::Tiny

    *   Role::Basic

    *   p5-mop-redux <https://github.com/stevan/p5-mop-redux>

  Functions
    `get_role_info($package_name)`
        Returns a hashref of information about a role; returns `undef` if the
        package does not appear to be a role. Attempts to load the package
        using Module::Runtime if it's not already loaded.

        The hashref may contain the following keys:

        *   `name` - the package name of the role

        *   `type` - the role implementation used by the role

        *   `api` - an arrayref of method names required/provided by the role

        *   `provides` and `requires` - the same as `api`, but split into
            lists of methods provided and required by the role

        *   `meta` - a metaobject for the role (e.g. a Moose::Meta::Role
            object). This key may be absent if the role implementation does
            not provide a metaobject

        This function may be exported, but is not exported by default. If you
        do not wish to export it, you may call it as a class method:

           Role::Inspector->get_role_info($package_name)

    `Role::Inspector::learn { BLOCK }`
        In the unlikely situation that you have to deal with some other role
        implementation that Role::Inspector doesn't know about, you can teach
        it:

           use Role::Inspector qw( learn );
   
           learn {
              my $r = shift;
              return unless My::Implementation::is_role($r);
              return {
                 name     => $r,
                 type     => 'My::Implementation',
                 provides => [ sort(@{My::Implementation::provides($r)}) ],
                 requires => [ sort(@{My::Implementation::requires($r)}) ],
              };
           };

        An alternative way to do this is:

           push @Role::Inspector::SCANNERS, sub {
              my $r = shift;
              ...;
           };

        You can do the `push` thing without having loaded Role::Inspector.
        This makes it suitable for doing inside My::Implementation itself,
        without introducing an additional dependency on Role::Inspector.

        Note that if you don't provide all of `provides`, `requires`, and
        `api`, Role::Inspector will attempt to guess the missing parts.

CAVEATS
    *   It is difficult to distinguish between Moo::Role and Role::Tiny roles.
        (The distinction is not often important anyway.) Thus sometimes the
        `type` for a Moo::Role may say "Role::Tiny".

    *   The way that Role::Basic roles are detected and introspected is a bit
        dodgy, relying on undocumented methods.

    *   Where Moose or Mouse roles define attributes, those attributes tend to
        result in accessor methods being generated. However neither of these
        frameworks provides a decent way of figuring out which accessor
        methods will result from composing the role with the class.

        Role::Inspector does its damnedest to figure out the list of likely
        methods, but (especially in the case of unusual attribute traits) may
        get things wrong from time to time.

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=Role-Inspector>.

SEE ALSO
    Class::Inspector.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2014 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
Role-Inspector
==============

Created:      2014-04-08
Home page:    <https://metacpan.org/release/Role-Inspector>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Role-Inspector>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.004	2014-04-15

 - Added: Include 'provides' and 'requires' keys in the returned hashref.

0.003	2014-04-09

 [ Test Suite ]
 - Tests added for Moo::Role roles when Moose is in memory.
 - Tests added for Role::Tiny and Moose::Role roles when Moo is in memory.

0.002	2014-04-09

 [ Documentation ]
 - Correct name for @Role::Inspector::SCANNERS in example.

0.001	2014-04-08	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "introspection for roles",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 0,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.015, CPAN::Meta::Converter version 2.120921",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "Role-Inspector",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        develop   => { recommends => { "Dist::Inkt" => 0.001 } },
                        runtime   => {
                                       requires => {
                                         "Exporter::Shiny" => 0,
                                         "Module::Runtime" => 0,
                                         "perl" => 5.006,
                                         "Scalar::Util" => 0,
                                       },
                                     },
                        test      => {
                                       recommends => {
                                                       "Moo::Role"   => "1.000000",
                                                       "Moose::Role" => "2.0600",
                                                       "mop::role"   => 0,
                                                       "Mouse::Role" => "1.00",
                                                       "Role::Basic" => 0,
                                                       "Role::Tiny"  => "1.000000",
                                                     },
                                       requires   => { "Test::Modern" => 0.006 },
                                     },
                      },
  "provides"       => {
                        "Role::Inspector" => { file => "lib/Role/Inspector.pm", version => 0.004 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => {
                                          web => "http://rt.cpan.org/Dist/Display.html?Queue=Role-Inspector",
                                        },
                        homepage     => "https://metacpan.org/release/Role-Inspector",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          url  => "git://github.com/tobyink/p5-role-inspector.git",
                                          web  => "https://github.com/tobyink/p5-role-inspector",
                                        },
                        X_identifier => "http://purl.org/NET/cpan-uri/dist/Role-Inspector/project",
                      },
  "version"        => 0.004,
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

